src/CommunityToolkit.Maui.Core/Views/Popup/PopupExtensions.android.cs

```diff
@@ -227,35 +165,98 @@ public static void SetSize(this Dialog dialog, in IPopup popup, in AView contain
+				{
+					var size = popup.Content.Measure(windowSize.Width / density, windowSize.Height / density);
+					realContentWidth = (int)context.ToPixels(size.Width);
+					realContentHeight = (int)context.ToPixels(size.Height);
+
+					if (double.IsNaN(popup.Content.Width))
+					{
+						realContentWidth = popup.HorizontalOptions == LayoutAlignment.Fill ? (int)windowSize.Width : realContentWidth;
+					}
+					if (double.IsNaN(popup.Content.Height))
+					{
+						realContentHeight = popup.VerticalOptions == LayoutAlignment.Fill ? (int)windowSize.Height : realContentHeight;
+					}
+				}
+				else
+				{
+					realContentWidth = (int)context.ToPixels(popup.Content.Width);
+					realContentHeight = (int)context.ToPixels(popup.Content.Height);
+				}
 			}
 			else
 			{
-				realContentWidth = (int)context.ToPixels(popup.Content.Width);
-				realContentHeight = (int)context.ToPixels(popup.Content.Height);
+				realWidth = (int)context.ToPixels(popup.Size.Width);
+				realHeight = (int)context.ToPixels(popup.Size.Height);
 			}
 
-			realWidth = realWidth is 0 ? realContentWidth : realWidth;
-			realHeight = realHeight is 0 ? realContentHeight : realHeight;
+			realWidth = Math.Min(realWidth is 0 ? realContentWidth : realWidth, (int)windowSize.Width);
+			realHeight = Math.Min(realHeight is 0 ? realContentHeight : realHeight, (int)windowSize.Height);
 
 			if (realHeight is 0 || realWidth is 0)
 			{
 				realWidth = (int)(context.Resources?.DisplayMetrics?.WidthPixels * 0.8 ?? throw new InvalidOperationException($"Unable to determine width. {nameof(context.Resources.DisplayMetrics)} cannot be null"));
 				realHeight = (int)(context.Resources?.DisplayMetrics?.HeightPixels * 0.6 ?? throw new InvalidOperationException($"Unable to determine height. {nameof(context.Resources.DisplayMetrics)} cannot be null"));
 			}
+
+			popup.Content.Measure(realWidth / density, realHeight / density);
+		}
+
+		static Size GetWindowSize(IWindowManager? windowManager, ViewGroup decorView)
+		{
+			ArgumentNullException.ThrowIfNull(windowManager);
+
+			int windowWidth;
+			int windowHeight;
+			int statusBarHeight;
+			int navigationBarHeight;
+
+			if (OperatingSystem.IsAndroidVersionAtLeast((int)BuildVersionCodes.R))
+			{
+				var windowMetrics = windowManager.CurrentWindowMetrics;
+				var windowInsets = windowMetrics.WindowInsets.GetInsetsIgnoringVisibility(WindowInsets.Type.SystemBars());
+				windowWidth = windowMetrics.Bounds.Width();
+				windowHeight = windowMetrics.Bounds.Height();
+				statusBarHeight = windowInsets.Top;
+				navigationBarHeight = windowHeight < windowWidth ? windowInsets.Left + windowInsets.Right : windowInsets.Bottom;
+			}
+			else if (windowManager.DefaultDisplay is null)
+			{
+				throw new InvalidOperationException($"{nameof(IWindowManager)}.{nameof(IWindowManager.DefaultDisplay)} cannot be null");
+			}
+			else
+			{
+				APoint realSize = new();
+				APoint displaySize = new();
+				ARect contentRect = new();
```

Comment:
Can we verify if those values will change? I'm quite sure that `displaySize` will not change during the application lifetime, so we can cache it and avoid calling the Android APIs every time, this will improve perf. a lot. If they still change, and we need to do the calls to `widowManager` and `decorView`, can we reuse the objects, instead of creating new ones every time?

Reply:
@pictos , Thank you for your review.
The variables realSize, displaySize and contentRect change with screen rotation.
Fixed not to create an object for each call by storing it in a static variable.